cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/DistributedProgramContainerModule.java

```diff
@@ -0,0 +1,258 @@
+
+  /**
+   * Defines the mode that the program container is getting executed.
+   */
+  public enum ClusterMode {
+    /**
+     * The mode that program containers run in the same cluster as the CDAP,
+     * hence can communicate directly with CDAP
+     */
+    ON_PREMISE,
+
+    /**
+     * The mode that program containers run in isolated cluster such that it cannot communicate with CDAP directly.
+     */
+    ISOLATED
+  }
+
+  private final CConfiguration cConf;
+  private final Configuration hConf;
+  private final ProgramRunId programRunId;
+  private final String instanceId;
+  private final ClusterMode clusterMode;
+  @Nullable
+  private final String principal;
+  @Nullable
+  private final ServiceAnnouncer serviceAnnouncer;
+
+  private DistributedProgramContainerModule(CConfiguration cConf, Configuration hConf, ProgramRunId programRunId,
+                                            String instanceId, ClusterMode clusterMode,
+                                            @Nullable String principal, @Nullable ServiceAnnouncer serviceAnnouncer) {
+    this.cConf = cConf;
+    this.hConf = hConf;
+    this.programRunId = programRunId;
+    this.instanceId = instanceId;
+    this.clusterMode = clusterMode;
+    this.principal = principal;
+    this.serviceAnnouncer = serviceAnnouncer;
+  }
+
+  @Override
+  protected void configure() {
+    String txClientId = generateClientId(programRunId, instanceId);
+    List<Module> modules = getCoreModules(programRunId.getParent(), txClientId);
+
+    AuthenticationContextModules authModules = new AuthenticationContextModules();
+    modules.add(principal == null
+                  ? authModules.getProgramContainerModule()
+                  : authModules.getProgramContainerModule(principal));
+
+    install(Modules.override(modules).with(new AbstractModule() {
+      @Override
+      protected void configure() {
+        // Overrides the LineageWriter, UsageWriter to write to TMS instead
+        bind(LineageWriter.class).to(MessagingLineageWriter.class);
+        bind(UsageWriter.class).to(MessagingUsageWriter.class);
+      }
+    }));
+  }
+
+  private List<Module> getCoreModules(final ProgramId programId, String txClientId) {
+    List<Module> modules = new ArrayList<>();
+
+    modules.add(new ConfigModule(cConf, hConf));
+    modules.add(new IOModule());
+    modules.add(new ZKClientModule());
+    modules.add(new MetricsClientRuntimeModule().getDistributedModules());
+    modules.add(new MessagingClientModule());
+    modules.add(new LocationRuntimeModule().getDistributedModules());
+    modules.add(new DiscoveryRuntimeModule().getDistributedModules());
+    modules.add(new DataFabricModules(txClientId).getDistributedModules());
+    modules.add(new DataSetsModules().getDistributedModules());
+    modules.add(new AuditModule().getDistributedModules());
+    modules.add(new NamespaceClientRuntimeModule().getDistributedModules());
```

Comment:
shouldn't we have different bindings for `ISOLATED` ClusterMode:
namespace client
datafabric
dataset
audit

Reply:
I guess we actually don’t need the mode in this module. Could push these down to the AbstractProgramTwillRunnable level